### ISSUE TITLE
After Update: net::ERR_CONTENT_LENGTH_MISMATCH when serving static files

### ISSUE THREAD

**Issue by mwegithub** [civetweb/civetweb#417]
Hi, 
I'm using civetweb in one of my side projects.with Visual Studio 2015 on Windows 10.
For the past six months the project worked without problems. When I recently tried to add websockets I ran into some issues (hang in close() when shutting down the server.

So I first updated my (older) copy of civetweb. by doing a fresh clone from GitHub.  This (thanks!) solved the problems with the hang while closing the server.

But now I get net::ERR_CONTENT_LENGTH_MISMATCH error reports (or "security" errors in IE) when I load some of my test pages. They include **static files** (jquery, bootstrap etc.) and when I refresh the browser a few times I always get the **content length mismatch error** reports for one or more .js or .css files. It seems that the number of bytes delivered by civetweb does not match the content-length header anymore...

This never happened with the civetweb version I had before (which was maybe 3 months old).
Can I do anything to track this down?

**Comment by mwegithub** [comment]
I tried to figure out which version caused this problem. I have not much experience with git so I first tried the GitHub Windows client and also GitKraken. Either I'm too dumb or it is really hard to 'check out' a specific commit or version or tag with these tools.

Anyway, I learned the git command line tool and how to check out specific commits with it. I figured out that the commit 

43d1d430185490035d897da23f6634fa30e664bf
Possible fix for #336 (3)

Is causing the problem on my systems (I tried two computers with IE/Chrome/FF). When I check out this version, Chrome immediately reports a content length mismatch when civetweb is serving static files. **Note that you need to disable the cache** to see this - else the browser cache may hide the problem. I always disable the cache in the Chrome DevTools.

When I go back to the commit before that:

c6efb738af9a569cfe6e5022756001d40deaeda4
Possible fix for #336 (2)

The problem is gone.
I can refresh dozens of times (without cache so full reload from civetweb) and no errors are reported. 

I will look at the sources next, trying to figure out what has changed and what may be causing the content length issue.

**Comment by bel2125** [comment]
I personally only use the C interface, not the C++ wrapper class. The C++ wrapper is not complete and only sometimes updated by some contributors. So I cannot provide an example for a class, only some C examples.

You took the most recent source from github, or some release version (from Sourceforge or a github release)?
A 32 or 64 bit build?
The content length error is reported for a HTTP, HTTPS, WS or WSS connection?
Is there an easy way I could reproduce it?

I recently changed some code for the data alignment for websocket packaged > ~120 bytes length. How long are the websocket requests you are using?

**Comment by mwegithub** [comment]
I use the master version (1.9) from github, building with Visual Studio 2015 on Windows 10, 32-Bit application. HTTP connection. Browser and civetweb hosting application on the same PC.

I worked on this last night and I can even reproduce it when I **don't** use web sockets.  So this rules out that.

When I press <F5> quickly enough in Chrome it reports issues like:

GET http://127.0.0.1:49776/system/jquery/dist/jquery.min.js net::ERR_CONTENT_LENGTH_MISMATCH

I can repro this with a simple HTML page with just a few included .js and .css files and an empty <body>.

I wonder if this is something specific caused by disabling the caching in the browsers (for debugging purposes). I don't see this problem (at least in my recent tests) when caching is enabled.

My next steps will be to log the file names and content lengths in civetweb. If they are correct, we know its a problem in the browsers. I will report back as soon as I know more.

**Comment by mwegithub** [comment]
Figured it out, I think.

I had a look at the source code of the _handle_static_file_request_ function. I added some debug traces, and this _changed the behavior. The error did not appear as often. So, some timing issue, maybe. After minimizing the debug output I could again repro the problem.

And, indeed, civitweb sent less bytes to the browser than specified content-length header. Further tracking this down via send_file_data() which stopped sending data because mg_write() returned -1. I finally found that the push() returned -1 because it timed out

(start - now) > timeout_ns

My guess now is that when a browser does not cache and a user press F5 often or just in the right spot or the page reloads itself, that the browser may try to handle multiple requests while aborting old, not yet answered requests. And this stalls something that causes send() in push() to take too long....?

When push() returns -1, the err is 0. It's only the timeout causing the "problem".
Sometimes the WSAEWOULDBLOCK is returned by send() in that case.

I think this is not a bug in civertweb but just cased by how the browsers behave when you rapidly reload pages while caching is disabled and a total reload of all static files is required. It's a timing issue and it changes after adding some (slow) calls to OutputDebugString into the _handle_static_file_request_ function_

**Comment by bel2125** [comment]
- So, it worked with 1.8 and does not with 1.9, and you are using HTTP (not HTTPS or websocket) in both cases?
- Did you test the most recent version from github as well - not a release, but the current source - the 1.10 "work in progress" version?
- Are you using only localhost (127.0.0.1)?

> I think this is not a bug in civertweb but just cased by how the browsers behave when you rapidly reload pages while caching is disabled and a total reload of all static files is required. It's a timing issue.

The -1 return for localhost means, the client is not reading the data much slower (or not at all) as the server can send it. What timeout value did you set in your configuration?

> Maybe civetweb should not return a header at all when push() fails and instead return a 400 or so?

That's not possible, since the first data civetweb (and any other HTTP server) will send is the header. So, there is no error before trying to send the header.  And if a "send" call later fails, the header has already been sent.

It seems, when you reload very often in your browser, the browser opens many connections and does not read fast enough from them, so the network buffers in windows itself overflow, leading to a "send failed" at the server. It's ok for the server to retry for a while, but finally it must close these connections sooner or later, because they block resources (ports, threads, memory) at the server.

There were some changes between 1.8 and 1.9 that could explain the differences, and some between 1.9 and the current 1.10 "working" that could change this again. But I'm not sure if this is a bug, or an unavoidable consequence of overloading the browser to a point it would block the server, and the server "defends" against overload by closing connections after some timeout.

**Comment by mwegithub** [comment]
I'm not really sure if this specific behavior is "new" to civetweb or if I just never discovered it before. I currently use civetweb to host a webservice and I use only local host connections over HTTP or HTTP. 

I use a 600s request timeout, and alll other timeouts are default. I did not see a change in the behavior with keep_allive on/off. I allow for 50 parallel threads.

I consider this to be a special 'behavior' that only shows up when we use the browser in developer mode (with caching turned off). In my (shallow) tests I could never see this behavior when the cache was enabled (the default). But I can repro it with a few <F5>'s in Chrome and Edge when caching is off. Waiting a second and then pressing <F5> again solves the problem always.

The "net" tabs in the browsers don't show anything specific, it's just the console that reports the wrong content length and the page of course fails to load.

**Comment by bel2125** [comment]
Could you attach the debug log? If you do a console build with DEBUG set, the server will log connects and disconnects.

What size is your test file?

Did you set the `static_file_max_age` settings?

**Comment by mwegithub** [comment]
I can repro this in Chrome easily after setting the "Disable Cache" option in the Net tab of the developer tools. I don't know which "debug log" you mean. I use the access_logfile and error_logfile configuration options.

When Chrome reports the problem (e.g. while trying to load a .css file) I can see in the access log that the content length returned by the GET request is indeed less than it is when the request is processed normally. For example:
OK: 
dist/css/bootstrap.min.css.map HTTP/1.1" 200 542194

FAIL:
dist/css/bootstrap.min.css.map HTTP/1.1" 200 368640
dist/css/bootstrap.min.css.map HTTP/1.1" 200 114688

This can happen with all static files. I can repro it reliably, but only when I use the Disable Cache option and press F5 quickly (to force a reload of static files changed in an editor). The error log is empty.

The push() function fails because the browser is already reloading other files or something. Surely not a bug in civetweb.  No reason to spend more time with this.

Thank you very much for your support!

**Comment by mwegithub** [comment]
I've just tried with the latest version (after reading the #422 thread).
My first tests show that the error is now gone, no more content length errors when F5ing a page in Chrome with disabled cache. This is **very good news indeed**.

Thank you for your support and excellent work on this project.

**Comment by photools** [comment]
A couple of days ago I creates a new clone from the current sources.
Now I'm seeing the dreaded ERR_CONTENT_LENGTH_MISMATCH error again. Is this a regression?

**Comment by photools** [comment]
It seems that the code block starting at line 5206

if (sret > 0) {
  /* We got ready to write. Don't check the timeout
  * but directly go back to write again. */
  continue;
}

is causing the issue. I've put a trace into this and whenever sret becomes > 0, Chrome reports the dreaded net::ERR_CONTENT_LENGTH_MISMATCH afterwards and my page fails to load.

When I comment out the continue instruction, the error is gone.

**Comment by bel2125** [comment]
This line was introduced in #474

I was not able to reproduce this issue in the first place, I doubt I can reproduce it now.
But I was not able to reproduce #474 either.
Maybe the temporary fix of just using [sleep](https://github.com/civetweb/civetweb/commit/f0733b76a8d46c2199f2fe3b2822b73de1981fe6#diff-942b175a4f5673b8f0ce6c0ab2fc2932R5080) instead of select works better? But I would not know why, and I can't investigate it since I can't reproduce it.

**Comment by bel2125** [comment]
No, I'm again not able to reproduce it. But I tested on Linux. (The Windows 10 machine needs some maintenance, and I won't get it back for a week).

A similar issue was due to a copy paste error ([here](https://github.com/civetweb/civetweb/issues/422#issuecomment-291978229)), but this is fixed now. 

But I have another idea:
When initializing `start = mg_get_current_time_ns();` (currently [line 5090](https://github.com/civetweb/civetweb/blob/master/src/civetweb.c#L5090)), initialize `now` as well:
`start = now = mg_get_current_time_ns();`

**Comment by photools** [comment]
Thank you for looking into this!

I have a 100% repro case using Chrome or an embedded Chromium-based browser.
After shipping my software on Saturday with the latest Civetweb version included, I've immediately received bug reports about the embedded browser "not rendering properly" or the net::ERR_CONTENT_LENGTH_MISMATCH error message showing in the Chrome console when they run tests with caching disabled.

I tried your second idea (start = now = ...)  first and **it worked!**
I've changed the code in line 5089 to 
```
if (timeout > 0) {
    start = now = mg_get_current_time_ns();
```
and this solves the problem nicely. I could not reproduce it anymore after making this change. Changing back immediately produced the problem again.

This seems to solve the issue again. Thank you very much.

Besides: I run a Visual Studio 2015 code analysis on the civetweb.c and it reported a few potential NULL pointer references or reading beyond buffers. If you don't have Visual C++ available I can attach the report and the version of the civetweb.c I used (so the line numbers match). May be worth looking over, not many issues reported.

**Comment by bel2125** [comment]
> If you don't have Visual C++ available I can attach the report and the version of the civetweb.c 
> I used (so the line numbers match). May be worth looking over, not many issues reported.

I do have Visual Studio available - from 2015 only the Express/Community version. I tend to use Coverity as a static source code analysis tool. Different static source code analysers tend to report different problems, often some of them are false positives.
But, please just post the warnings you get here, I will have a look.

**Comment by photools** [comment]
[ca--civ-05.zip](https://github.com/civetweb/civetweb/files/1155610/ca--civ-05.zip)

To keep things simple I've just made screen shots of the corresponding source annotations and the error message and details created by VC code analysis. I compiled using VC 2015 (SP 3) for the 64-bit platform.

There are two additional warnings about data size mismatches for two DEBUG_TRACE calls where 64-bit integers used as parameters but the format string uses %d (32-bit int):

15253: DEBUG_TRACE("grabbed socket %d, going busy", sp ? sp->sock : -1);
and
15290: DEBUG_TRACE("grabbed socket %d, going busy", sp ? sp->sock : -1);

Have a look if you have a minute. Maybe it can improve the code a bit or make it more safe. 
Thanks again for your awesome work on Civetweb.
